### PR TITLE
Support/wagtail 60

### PR DESCRIPTION
**Pull request by katdom13**
[Support Ticket](https://torchbox.monday.com/boards/1124794299/pulses/1396934613)

### Summary of changes

- Added Wagtail 6.0 compatibility
- Added tests for Django 5.0
- Added tests for Python 3.12

### CODE REVIEW

**Comment by katdom13** [comment]
Unfortunately,
I couldn't tweak this one:
https://github.com/torchbox-forks/wagtail-footnotes/blob/main/wagtail_footnotes/widgets.py#L5

Based on this release note:
https://docs.wagtail.org/en/stable/releases/6.0.html#deprecated-widgetwithscript-base-widget-class

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -72,18 +72,18 @@ jobs:
...
+      - uses: actions/setup-python@v5
         with:
           # Use latest Python, so it understands all syntax.
           python-version: ${{env.PYTHON_LATEST}}
 
       - run: python -Im pip install --upgrade "coverage[toml]>=7.2,<8.0"
 
       - name: ⬇️ Download coverage data
-        uses: actions/download-artifact@v3
+        uses: actions/download-artifact@v4
```
Comment (zerolab) [inline .github/workflows/test.yml:86]: best to keep this and `upload-artifact` to v3, or follow the instructions from https://hynek.me/articles/ditch-codecov-python/

File: pyproject.toml [hunk 2]
```diff
@@ -23,14 +23,13 @@ classifiers = [
...
     "Programming Language :: Python :: 3.12",
     "Framework :: Django",
-    "Framework :: Django :: 3",
     "Framework :: Django :: 3.2",
-    "Framework :: Django :: 4",
-    "Framework :: Django :: 4.1",
     "Framework :: Django :: 4.2",
+    "Framework :: Django :: 5.0",
     "Framework :: Wagtail",
     "Framework :: Wagtail :: 4",
```
Comment (zerolab) [inline pyproject.toml]: now that 4.x is out of support, let's go with `Wagtail >= 5.2`, so

```suggestion
```

File: tests/settings.py [hunk 3]
```diff
@@ -38,7 +40,7 @@
     "wagtail.images",
     "wagtail.search",
     "wagtail.admin",
-    "wagtail.contrib.modeladmin",
+    "wagtail_modeladmin" if WAGTAIL_VERSION >= (5, 1) else "wagtail.contrib.modeladmin",
```
Comment (zerolab) [inline tests/settings.py]: by dropping support for Wagtail < 5.2, we no longer need this

File: tox.ini [hunk 4]
```diff
@@ -28,16 +28,17 @@ deps =
...
 
     wagtail4.1: wagtail>=4.1,<4.2
-    wagtail5.1: wagtail>=5.1,<5.2
     wagtail5.2: wagtail>=5.2,<5.3
+    wagtail6.0: wagtail>=6.0,<6.1
 
 
 install_command = python -Im pip install -U {opts} {packages}
 commands =
+    python -Im pip install -e '.[testing]' -U
```
Comment (zerolab) [inline tox.ini]: please move this to the `extras` config https://tox.wiki/en/4.13.0/config.html#extras

```ini
deps =
  ...

extras = testing
```

File: wagtail_footnotes/widgets.py [hunk 5]
```diff
@@ -1,20 +1,66 @@
...
+            return mark_safe(shown + hidden)  # noqa: S308
+
+        def render(self, name, value, attrs=None, renderer=None):
+            # Ensure 'id' is present in attrs
+            try:
+                attrs["id"]
+            except (KeyError, TypeError):
+                raise TypeError(  # noqa: B904
+                    "ReadonlyUUIDInput cannot be rendered without an 'id' attribute"
+                )
```
Comment (zerolab) [inline wagtail_footnotes/widgets.py]:
```suggestion
            except (KeyError, TypeError) as exc:
                raise TypeError(
                    "ReadonlyUUIDInput cannot be rendered without an 'id' attribute"
                ) from exc
```

File: wagtail_footnotes/widgets.py [hunk 6]
```diff
@@ -1,20 +1,66 @@
...
+
+            # Render the HTML portion
+            widget_html = self.render_html(name, value, attrs)
+
+            # Combine HTML and JavaScript, and mark as safe
+            out = f"{widget_html}<script></script>"
```
Comment (zerolab) [inline wagtail_footnotes/widgets.py]: I am not entirely convinced we need the empty script tag

**Review by zerolab — CHANGES_REQUESTED** [review]
Left a number of notes.

My main gripe is: why do we have a PR on a fork for a package that we have control of and maintain? i.e. this PR should be against https://github.com/torchbox/wagtail-footnotes/

File: .github/workflows/test.yml (hunk 1 above)
Comment (katdom13) [reply .github/workflows/test.yml:86]: Hi @zerolab ,
Thank you for the resource! I pushed forward with the version upgrade.
https://github.com/torchbox-forks/wagtail-footnotes/pull/2/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R66

File: pyproject.toml (hunk 2 above)
Comment (katdom13) [reply pyproject.toml]: Removed this here, thanks:
https://github.com/torchbox-forks/wagtail-footnotes/pull/2/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711L32

File: tests/settings.py (hunk 3 above)
Comment (katdom13) [reply tests/settings.py]: Removed this one too:
https://github.com/torchbox-forks/wagtail-footnotes/pull/2/files#diff-cf01e5d691c05467162c955fc1688cc98afdb2ad6c6640caacb2569b0735a7c7L41

As well as this conditional:
https://github.com/torchbox-forks/wagtail-footnotes/pull/2/files#diff-30fbe45aee5bfdf1115f199ae065d43517529fa8ce35b8edf209b6d409c7cee6L57

File: tox.ini (hunk 4 above)
Comment (katdom13) [reply tox.ini]: moved here:
https://github.com/torchbox-forks/wagtail-footnotes/pull/2/files#diff-ef2cef9f88b4fe09ca3082140e67f5ad34fb65fb6e228f119d3812261ae51449R39

File: wagtail_footnotes/widgets.py (hunk 5 above)
Comment (katdom13) [reply wagtail_footnotes/widgets.py]: Thanks! Applied here:
https://github.com/torchbox-forks/wagtail-footnotes/pull/2/files#diff-ac21e48aad8191b1f8c978e106163cbad7084695f6f812fafc6b0cc30912e1b9R20-R23

File: wagtail_footnotes/widgets.py (hunk 6 above)
Comment (katdom13) [reply wagtail_footnotes/widgets.py]: With the deprecation of `WidgetWithScript` in Wagtail 6.0, I tried converting this to stimulus, but failed to do so. A workaround I did for now was to remove `WidgetWithScript`, but copy its `render` method:
https://github.com/torchbox-forks/wagtail-footnotes/pull/2/files#diff-ac21e48aad8191b1f8c978e106163cbad7084695f6f812fafc6b0cc30912e1b9R15

**Comment by katdom13** [comment]
> Left a number of notes.
> 
> My main gripe is: why do we have a PR on a fork for a package that we have control of and maintain? i.e. this PR should be against https://github.com/torchbox/wagtail-footnotes/

A rationale for this was because of a past discussion and agreement about our support workflow.

The current workflow is:
1. Do the package upgrade work in your work branch
2. Create a PR on the forks repo
3. A support member will code review
4. If approved, merge to the forked repo's `main`/`master` branch
5. Create a PR against the maintainer's repo, the source being our forked repo's `main`/`master` branch

This applied to packages regardless of who the maintainer is.

Though I'm happy to create another PR directly to the maintainer repo.
I opened this discussion on our private support channel:
https://torchbox.slack.com/archives/C03DHB7TJQH/p1708413847263609

Please see the PR to the maintainer's repo here:
https://github.com/torchbox/wagtail-footnotes/pull/66


Thanks for your insights!

**Comment by zerolab** [comment]
Thanks @katdom13,

> A rationale for this was because of a past discussion and agreement about our support workflow.

It is possible I missed this at TNM, but as a maintainer, I do not wish to review things on a fork, then re-review things on the proper repo.

> The current workflow is:
> 
>     1. Do the package upgrade work in your work branch
>     2. Create a PR on the forks repo
>     3. A support member will code review
>     4. If approved, merge to the forked repo's `main`/`master` branch
>     5. Create a PR against the maintainer's repo, the source being our forked repo's `main`/`master` branch
>
> 
> This applied to packages regardless of who the maintainer is.

This doubles the work for packages under our control. Just open the PR directly. The support member can do the initial review, then the maintainer (which will be someone in our team), can then do the final review. Whether the PR comes from your fork, or the torchbox-forks version, it shouldn't matter.

 
> Though I'm happy to create another PR directly to the maintainer repo. I opened this discussion on our private support channel: https://torchbox.slack.com/archives/C03DHB7TJQH/p1708413847263609

I don't have access to it, so can't pitch in :)

**Comment by nickmoreton** [comment]
Hi @zerolab Yep I agree having to code review twice isn't a good plan, I don't think we intended that.

Over the last year or so we found ourselves trying to manage around 40+ wagtail packages. Some we can push through and avoid blocking the upgrade work but most don't get a new release before the upgrade needs to be deployed. We then manage our own repo tags off our main branches and use them as the dependencies.

We came up with the idea to fork every repo in our own org so that we don't use our own personal accounts as there are so many to deal with, then as the staff in support change over time whoever is working in support can pick up where others have left off. There's info about it here: https://intranet.torchbox.com/torchbox-teams/tech-team/working-with-3rd-party-packages/

It's not easy for use to "remember" packages that need to be dealt with differently such as this one so we decided to follow the same process for any package used in a client site that we need to upgrade as a hosting and support task.

We'd be happy to change the process if you think we should.

**Comment by zerolab** [comment]
@nickmoreton and @katdom13 
I appreciate the effort the team puts into these.

My goal is to reduce effort duplication for the packages that we maintain (outside of the core team). If you end up creating pseudo-releases by having worked merged and tagged on a fork for a package and repository that we have full control of, then I feel we're doing it wrong. Our goal, as a team is to do better at maintaining and making new releases for the packages we oversee. As such, how about the support team become co-maintainers on the packages that historically have taken too long to review PRs and make releases. That way we can make releases early and often and everyone benefits?

**Comment by nickmoreton** [comment]
@zerolab @katdom13 

> As such, how about the support team become co-maintainers on the packages that historically have taken too long to review PRs and make releases. That way we can make releases early and often and everyone benefits?

Sounds perfect to me and Im sure the rest of the team will agree it's a great step forward.

**Pull request merged**